File: packages/nodx-py/src/nodx/canonical.py
import json


def canonical_json(doc):
    return canonical_stringify(
        {"body": doc["body"], "meta": doc["meta"], "schema": doc["schema"]}
    )


def canonical_stringify(value):
    return json.dumps(sort_value(value), ensure_ascii=False, separators=(",", ":"))


def sort_value(value):
    if isinstance(value, list):
        return [sort_value(item) for item in value]
    if isinstance(value, dict):
        return {key: sort_value(value[key]) for key in sorted(value)}
    return value

File: packages/nodx-js/src/canonical.mjs
export function canonicalJson(doc) {
  const { body, meta, schema } = doc;
  return canonicalStringify({ body, meta, schema });
}

export function canonicalStringify(value) {
  return JSON.stringify(sortValue(value));
}

export function sortValue(value) {
  if (Array.isArray(value)) return value.map(sortValue);
  if (value && typeof value === "object") {
    return Object.fromEntries(Object.keys(value).sort().map((key) => [key, sortValue(value[key])]));
  }
  return value;
}
